src/types/connection/connection_type.rs

```diff
@@ -1,97 +1,136 @@
+use crate::connection::edge::Edge;
+use crate::connection::page_info::PageInfo;
+use crate::types::connection::{CursorType, EmptyEdgeFields};
 use crate::{
-    do_resolve, registry, Context, ContextSelectionSet, EmptyEdgeFields, Error, ObjectType,
+    do_resolve, registry, Context, ContextSelectionSet, Error, FieldResult, ObjectType,
     OutputValueType, Positioned, QueryError, Result, Type,
 };
 use async_graphql_parser::query::Field;
+use futures::{Stream, TryStreamExt};
 use indexmap::map::IndexMap;
 use inflector::Inflector;
 use itertools::Itertools;
 use std::borrow::Cow;
 
+/// The record type output by the data source
+pub struct Record<C, E, T> {
+    cursor: C,
+    edge: E,
+    element: T,
+}
+
+impl<C, E, T> Record<C, E, T> {
+    pub fn new(cursor: C, edge: E, element: T) -> Self {
+        Self {
+            cursor,
+            edge,
+            element,
+        }
+    }
+}
+
+impl<C, T> Record<C, EmptyEdgeFields, T> {
+    pub fn new_without_edge_fields(cursor: C, element: T) -> Self {
+        Self {
+            cursor,
+            edge: EmptyEdgeFields,
+            element,
+        }
+    }
+}
+
 /// Connection type
 ///
-/// Connection is the result of a query for `DataSource`,
-/// If the `T` type is `OutputValueType`, you can return the value as a field function directly,
-/// otherwise you can use the `Connection::map` function to convert to a type that implements `OutputValueType`.
-/// `E` is an extension object type that extends the edge fields.
+/// Connection is the result of a query for `DataSource`.
 pub struct Connection<T, E: ObjectType + Send = EmptyEdgeFields> {
     /// The total number of records.
-    pub total_count: Option<usize>,
+    pub(crate) total_count: Option<usize>,
 
     /// Information about pagination in a connection.
-    pub page_info: PageInfo,
+    pub(crate) page_info: PageInfo,
 
     /// All records of the current page.
-    pub nodes: Vec<(Cursor, E, T)>,
+    pub(crate) nodes: Vec<Record<String, E, T>>,
 }
 
-impl<T, E: ObjectType + Send> Connection<T, E> {
-    /// Create a connection object.
-    pub fn new(
-        total_count: Option<usize>,
+impl<T: OutputValueType + Send, E: ObjectType + Send> Connection<T, E> {
+    pub fn empty() -> Self {
+        Connection {
+            total_count: None,
+            page_info: PageInfo {
+                has_previous_page: false,
+                has_next_page: false,
+                start_cursor: None,
+                end_cursor: None,
+            },
+            nodes: Vec::new(),
+        }
+    }
+
+    pub fn new<C>(
+        nodes: Vec<Record<C, E, T>>,
         has_previous_page: bool,
         has_next_page: bool,
-        nodes: Vec<(Cursor, E, T)>,
-    ) -> Self {
+        total_count: Option<usize>,
+    ) -> Connection<T, E>
+    where
+        C: CursorType,
+    {
+        let nodes = nodes
+            .into_iter()
+            .map(|record| Record {
+                cursor: record.cursor.encode_cursor(),
```

Comment:
I'm wondering if this needs to be similar to `try_into` since it should be allowed to fail.